### ISSUE TITLE
Is there any way to slim down the r-devel image?

### ISSUE THREAD

**Issue by shabbychef** [rocker-org/rocker#169]
By my accounting, the `r-devel` image is bigger than 3G: 
```
$ docker images | grep r-devel
rocker/r-devel                     latest              0c2a04c8e39a        3 months ago        3.214 GB
```
Would it be possible to slim this down, possibly by removing packages that were installed to build R?
For this to have a meaningful difference to download times, the `RUN` commands would have to be merged into one, I think, which is awful for development, but better for the end user.

**Comment by eddelbuettel** [comment]
Are you aware of [drd](https://hub.docker.com/r/rocker/drd/) ? 

> drd: Docker R-devel
This repository aims to provide (possibly daily) builds of Docker containers based on
the then-current state of the R-devel SVN repository.

> It is now part of the rocker
repository in which @cboettig and @eddelbuettel are now working, and it
builds upon earlier Docker repos and automated builds in Dirk's repositories.

> So what about r-devel in Rocker then?
Excellent question, but the answer will get nerdy quickly. Docker containers
are composed of layers implemented via AUFS.
Each RUN command is one layer. This makes it impossible to reclaim
filespace from an earlier layer in a later layer -- so the R-devel container
can not easily remove (temp.) files and directories.

> For drd we went the other route: The container is built in a single
invocation of RUN which permits us to reclaim space for an overall smaller
container. But it is sort-of running against the grain of how Docker
containers are built, and hence still somewhat experimental.

> But if you just want a smaller container with a current R-devel build, this
may be the one.

**Comment by eddelbuettel** [comment]
But looks like it got bigger too:

```bash
edd@max:~$ docker images | grep "drd\|r-devel "
rocker/drd                   latest                14c7d2cdae28        19 hours ago         2.241 GB
rocker/r-devel               latest                8120d0a9b459        12 days ago          3.369 GB
rocker/drd                   master                cd5f99f62b14        13 months ago        820.4 MB
edd@max:~$ 
```

**Comment by shabbychef** [comment]
`drd` looks exactly like what I wanted. I would suspect that an end user would _always_ prefer `drd` to `r-devel` except for the possibility that the R-devel SVN directory would contain a broken version of R-devel, whereas `r-devel` is presumably more stable? Is that right?

**Comment by eddelbuettel** [comment]
No I think they start from the same spot. 

It's been a while but I think I meant r-devel to be use to develop,  ie more Dev packages etc,  and drd just to deploy.

**Comment by shabbychef** [comment]
OK, with that advice, I'll close this and migrate to `drd`. (Strangely, docker-hub reports it as 901MB, but it shows up as 2.24 GB on my machine after a pull.)

**Comment by eddelbuettel** [comment]
We need to look at it -- it used to be an ~800mb image.

**Comment by shabbychef** [comment]
I had some time, so I broke up the `drd` Dockerfile into separate `RUN` commands, and then used `dockviz` to look at the tree of images. Here is the annotated Dockerfile, with the incremental and cumulative image sizes at completion of each `RUN` in comments after the `RUN`:
```
## Emacs, make this -*- mode: sh; -*-

## start with the Docker 'base R' Debian-based image
FROM r-base:latest
# │                     └─90d85b98acb7 Incremental Size: 0.0 B Tags: r-base:latest
# │                     └─90d85b98acb7 Cumulative Size: 1.0 GB Tags: r-base:latest

## That's me
## err, you.
MAINTAINER Dirk Eddelbuettel edd@debian.org
# │                       └─705bb9ad939b Incremental Size: 0.0 B
# │                       └─705bb9ad939b Cumulative Size: 1.0 GB

## Remain current
RUN apt-get update -qq && apt-get dist-upgrade -y
# │                         └─ff5b455320d8 Incremental Size: 776.8 MB
# │                         └─ff5b455320d8 Cumulative Size: 1.8 GB

## From the Build-Depends of the Debian R package, plus subversion
## Check out R-devel
## Build and install according the standard 'recipe' I emailed/posted years ago
## Set Renviron.site to get libs from base R install
## Clean up
## -- all in one command to get a single AUFS layer
RUN apt-get update -qq 
# │                           └─349e83a5fd06 Incremental Size: 17.1 MB
# │                           └─349e83a5fd06 Cumulative Size: 1.8 GB

RUN apt-get install -y -t unstable --no-install-recommends \
								bash-completion \
								bison \
								debhelper \
								default-jdk \
								g++ \
								gcc \
								gfortran \
								groff-base \
								libblas-dev \
								libbz2-dev \
								libcairo2-dev/unstable \
								libcurl4-openssl-dev/unstable \
								libfreetype6-dev/unstable \
								libharfbuzz-dev/unstable \
								libjpeg-dev \
								liblapack-dev \
								liblzma-dev \
								libncurses5-dev \
								libpango1.0-dev/unstable \
								libpcre3-dev \
								libpng-dev \
								libreadline-dev \
								libtiff5-dev/unstable \
								libx11-dev \
								libxcb1-dev/unstable \
								libxdmcp-dev/unstable \
								libxt-dev \
								mpack \
								subversion \ 
								tcl8.6-dev \
								texinfo \
								texlive-base \
								texlive-fonts-recommended \
								texlive-generic-recommended \
								texlive-latex-base \
								texlive-latex-recommended \
								tk8.6-dev \
								x11proto-core-dev \
								xauth \
								xdg-utils \
								xfonts-base \
								xvfb \
								zlib1g-dev 
# │                             └─f0dcf4a15a09 Incremental Size: 638.5 MB
# │                             └─f0dcf4a15a09 Cumulative Size: 2.5 GB

RUN cd /tmp \
    && svn co http://svn.r-project.org/R/trunk R-devel 
# │                               └─68706a23aab8 Incremental Size: 126.3 MB
# │                               └─68706a23aab8 Cumulative Size: 2.6 GB

RUN cd /tmp/R-devel && \
								R_PAPERSIZE=letter \
								R_BATCHSAVE="--no-save --no-restore" \
								R_BROWSER=xdg-open \
								PAGER=/usr/bin/pager \
								PERL=/usr/bin/perl \
								R_UNZIPCMD=/usr/bin/unzip \
								R_ZIPCMD=/usr/bin/zip \
								R_PRINTCMD=/usr/bin/lpr \
								LIBnn=lib \
								AWK=/usr/bin/awk \
								CFLAGS=$(R CMD config CFLAGS) \
								CXXFLAGS=$(R CMD config CXXFLAGS) \
								./configure --enable-R-shlib \
												--without-blas \
												--without-lapack \
												--with-readline \
												--without-recommended-packages \
												--program-suffix=dev 
# │                                 └─3b48ff7f5e46 Incremental Size: 1.2 MB
# │                                 └─3b48ff7f5e46 Cumulative Size: 2.6 GB

RUN cd /tmp/R-devel && \
							make && \
							make install && \
							rm -rf /tmp/R-devel /tmp/downloaded_packages/ /tmp/*.rds 
# │                                   └─c408d3185341 Incremental Size: 53.6 MB
# │                                   └─c408d3185341 Cumulative Size: 2.6 GB

RUN echo "R_LIBS=\${R_LIBS-'/usr/local/lib/R/site-library:/usr/local/lib/R/library:/usr/lib/R/library'}" >> /usr/local/lib/R/etc/Renviron 
# │                                     └─04c97b899c3a Incremental Size: 1.6 KB
# │                                     └─04c97b899c3a Cumulative Size: 2.6 GB

RUN echo 'options("repos"="https://cran.rstudio.com")' >> /usr/local/lib/R/etc/Rprofile.site 
# │                                       └─535183354b51 Incremental Size: 44.0 B
# │                                       └─535183354b51 Cumulative Size: 2.6 GB

RUN cd /usr/local/bin \
				&& mv R Rdevel \
				&& mv Rscript Rscriptdevel \
				&& ln -s Rdevel RD \
				&& ln -s Rscriptdevel RDscript 
# │                                         └─5f0afdfa2dd2 Incremental Size: 34.9 KB
# │                                         └─5f0afdfa2dd2 Cumulative Size: 2.6 GB

RUN dpkg --purge  \
								libblas-dev \
								libbz2-dev  \
								libcairo2-dev \
								libfontconfig1-dev \
								libfreetype6-dev \
								libglib2.0-dev \
								libharfbuzz-dev \
								libjpeg-dev \
								liblapack-dev  \
								liblzma-dev \
								libncurses5-dev \
								libpango1.0-dev \
								libpcre3-dev \
								libpng12-dev \
								libreadline-dev \
								libtiff5-dev \
								libxft-dev \
								r-base-dev \
								tcl8.6-dev \
								texlive-base \
								texlive-fonts-recommended \
								texlive-generic-recommended \
								texlive-latex-base \
								texlive-latex-recommended \
								tk8.6-dev 
# │                                           └─e89f5ca3b916 Incremental Size: 1.8 MB
# │                                           └─e89f5ca3b916 Cumulative Size: 2.7 GB

RUN apt-get autoremove -qy
# │                                             └─b25e0401e677 Incremental Size: 1.8 MB Tags: shabbychef/drd:latest
# │                                             └─b25e0401e677 Cumulative Size: 2.7 GB Tags: shabbychef/drd:latest
        
```

The main takeaway is that the `apt-get dist-upgrade` is huge.

This is an imperfect test because the cumulative sizes are apparently always monotonic increasing (confirmed by looking at output of all my images), so you don't see the benefit here from the `dpkg --purge`. Presumably it is 'only' on the order of 0.5 GB.

**Comment by shabbychef** [comment]
If I mostly collapse the commands back again in a Dockefile, you get the following tree of incremental sizes:
```
│                     └─90d85b98acb7 Virtual Size: 0.0 B Tags: r-base:latest
│                       ├─c6fda2a7a855 Virtual Size: 0.0 B
│                         └─2075cb25189d Virtual Size: 812.8 MB
│                           └─600c23f20a0c Virtual Size: 428.7 MB Tags: shabbychef/drd:latest
```
Again, the `dist-upgrade` is the `2075cb` image, with a bump of 812MB; the remaining operations add 'only' 428 MB, which I find surprisingly large.

So maybe the solution is for r-base to start from a fresher version of debian or ubuntu?

**Comment by eddelbuettel** [comment]
I really appreciate you digging through this, and I may have my head elsewhere: I don't quite see yet where it is coming from/  

Are you suggesting to change [this line](https://github.com/rocker-org/drd/blob/master/Dockerfile#L4) ?

All these seem similar:

```
edd@max:~/git$ docker images rocker/r-base
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
rocker/r-base       latest              e70915135bfa        5 days ago          1.076 GB
edd@max:~/git$ docker images r-base
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
r-base              latest              8b1651d59882        3 weeks ago         1.035 GB
r-base              3.2.3               7939bc63acf9        11 weeks ago        1.071 GB
r-base              3.2.2               d8d772db8b82        4 months ago        998.9 MB
r-base              3.2.1               38a6a6b62751        8 months ago        563.5 MB
r-base              3.2.0               c09b226d89bc        9 months ago        551.6 MB
r-base              3.1.3               8a0dd612a7f5        11 months ago       516.7 MB
r-base              3.1.2               c04052b3ddd2        12 months ago       483.5 MB
edd@max:~/git$ 
```

**Comment by cboettig** [comment]
> So maybe the solution is for r-base to start from a fresher version of debian or ubuntu? 

r-base is based on `debian:testing`... what did you have in mind?  

the Docker official repo folks would have preferred an older version I believe, you'll see many official images built on Jessie or Wheezy, but then they also don't like running `upgrade` or `dist-upgrade` in official Dockerfiles... Still, I'm surprised the upgrade makes such a difference; in theory those upgraded package versions should already be pulled into the lower-level images when they re-run the apt-get install commands...

**Comment by shabbychef** [comment]
Sorry, I think by 'fresher', I meant one which is idempotent with respect to `apt-get dist-upgrade`. I do not know how often the debian docker images are rebuilt, so maybe there is no such thing. This would have to be done in `r-base`. BTW, from Dirk's image listing, it seems there has been size creep over time of `r-base` as well, presumably due to a similar effect?

**Comment by cboettig** [comment]
@shabbychef good questions. 

Re: how frequently debian image is rebuilt, see: https://github.com/tianon/docker-brew-debian/issues/10#issuecomment-59330078 (answer: frequently enough).

 Right, r-base has been growing, and notably there's no `upgrade` line in r-base Dockerfile so that's not the issue. I guess this implies that the dependencies of R have just gotten significantly bigger/heavier; `debian-testing` remains around 118 mb...  Given that, I don't think changing the FROM source for the r-base image could save weight on either r-base or drd....

**Comment by eddelbuettel** [comment]
I think the best answer is to just sell your primary residence and move to wherever Google fiber is.

**Comment by shabbychef** [comment]
@eddelbuettel one step ahead of you: we rent.

**Comment by shabbychef** [comment]
By catting together all commands from the `Dockerfile`s for `r-base` and `drd`, I get basically no space savings--the image is still huge. I do not know much about Debian (_Ubuntu_ is Swahili for "Debian is too hard to configure."), but I am curious about the 'pinning'. In the `Dockerfile` for `r-base`, I see:
```
FROM debian:testing

...

RUN echo "deb http://http.debian.net/debian sid main" > /etc/apt/sources.list.d/debian-unstable.list \
	&& echo 'APT::Default-Release "testing";' > /etc/apt/apt.conf.d/default
```
Later in `drd`, there is the `apt-get dist-upgrade`. Could it possibly be the case that by _starting_ from `testing`, then pinning to `unstable` that the `dist-upgrade` is upgrading to `unstable`? Or did I misunderstand? Because I would guess that might make the upgrade large (like 800MB large).

**Comment by eddelbuettel** [comment]
That is possible.  But in general for package foo_1.2-3 from testing the package foo_1.2-4 from unstable should have roughly the same size.  Even foo_2.0-1 is rarely much bigger.

We also set up pinning to only go to unstable 'when told to do so', ie `apt-get install -t unstable ...`.  Of course, we may have gotten that wrong, or it may have changed.

I think the next will be running `du` over a few directory trees to see where on earth all that fluff hides.  If it is identifiable we can delete it away.   If not, time for more drinks...

**Comment by shabbychef** [comment]
I don't think `foo_1.2-4` will be much _bigger_ than `foo_1.2-3`, the problem is that an install of `foo_1.2-4` over `1.2-3` is the _entire_ size of the `1.2-4` install. So say that the `r-base` final image is 852MB, and is based on `debian:testing`, which is around 120MB; then if (and I don't know if this is true) the `dist-upgrade` is from `unstable`, you might see a delta on the order of 730MB or more, if _everything_ gets slight version bumps.
I guess one way to test this would be to start `r-base` `FROM debian:sid`. I'll try to run that and check the sizes.

**Comment by eddelbuettel** [comment]
Ping should prevent the wholesale upgrade, but what you suggest is good.   We could also re-order the commands to add `unstable` later (or not at all--I think we once needed it when Debian was frozen for a longish time prior to a release, things are again nimbler now).

**Comment by cboettig** [comment]
Sounds good to me too.  @shabbychef thanks for digging in here.  @eddelbuettel yup, might simplify things for downstream users if we were able to shed the need for apt-pinning to `unstable`.  (though in the short term might break dockerfiles that build on `r-base` & friends and use the `-t unstable`.

**Comment by shabbychef** [comment]
I had to cut out some lines because I got errors of the form:
```
The value 'unstable' is invalid for APT::Default-Release as such a release is not available in the sources
```
I removed the `dist-upgrade` in particular and any of the `-t unstable` flags for the same reason.

Here, then, is a merged `r-base` and `drd` Dockerfile annotated with the incremental and total image sizes:
```
## Emacs, make this -*- mode: sh; -*-
 
#FROM debian:testing
FROM debian:sid
#│ └─7b7dd7de6b48 Incremental Size: 0.0 B Tags: debian:sid
#│ └─7b7dd7de6b48 Total Size: 119.0 MB Tags: debian:sid

## This handle reaches Carl and Dirk
MAINTAINER "Carl Boettiger and Dirk Eddelbuettel" rocker-maintainers@eddelbuettel.com
#│   └─4bf263f317d3 Incremental Size: 0.0 B
#│   └─4bf263f317d3 Total Size: 119.0 MB

## Set a default user. Available via runtime flag `--user docker` 
## Add user to 'staff' group, granting them write privileges to /usr/local/lib/R/site.library
## User should also have & own a home directory (for rstudio or linked volumes to work properly). 
RUN useradd docker \
	&& mkdir /home/docker \
	&& chown docker:docker /home/docker \
	&& addgroup docker staff
#│     └─82def19dae26 Incremental Size: 330.9 KB
#│     └─82def19dae26 Total Size: 119.3 MB

RUN apt-get update \ 
	&& apt-get install -y --no-install-recommends \
		ed \
		less \
		locales \
		vim-tiny \
		wget \
		ca-certificates \
	&& rm -rf /var/lib/apt/lists/*
#│       └─12f907691ad8 Incremental Size: 63.1 MB
#│       └─12f907691ad8 Total Size: 182.4 MB

## Configure default locale, see https://github.com/rocker-org/rocker/issues/19
RUN echo "en_US.UTF-8 UTF-8" >> /etc/locale.gen \
	&& locale-gen en_US.utf8 \
	&& /usr/sbin/update-locale LANG=en_US.UTF-8
#│         └─108df614b04e Incremental Size: 1.6 MB
#│         └─108df614b04e Total Size: 184.1 MB

ENV LC_ALL en_US.UTF-8
#│           └─315e7760f948 Incremental Size: 0.0 B
#│           └─315e7760f948 Total Size: 184.1 MB
ENV LANG en_US.UTF-8
#│             └─3016e9b370db Incremental Size: 0.0 B
#│             └─3016e9b370db Total Size: 184.1 MB

## Use Debian unstable via pinning -- new style via APT::Default-Release
#RUN echo "deb http://http.debian.net/debian sid main" > /etc/apt/sources.list.d/debian-unstable.list \
	#&& echo 'APT::Default-Release "testing";' > /etc/apt/apt.conf.d/default

ENV R_BASE_VERSION 3.2.4
#│               ├─2e192b658a75 Incremental Size: 0.0 B
#│               ├─2e192b658a75 Total Size: 184.1 MB

## Now install R and littler, and create a link for littler in /usr/local/bin
## Also set a default CRAN repo, and make sure littler knows about it too
RUN apt-get update \
	&& apt-get install -y --no-install-recommends \
		littler \
                r-cran-littler \
		r-base=${R_BASE_VERSION}* \
		r-base-dev=${R_BASE_VERSION}* \
		r-recommended=${R_BASE_VERSION}* \
        && echo 'options(repos = c(CRAN = "https://cran.rstudio.com/"), download.file.method = "libcurl")' >> /etc/R/Rprofile.site \
        && echo 'source("/etc/R/Rprofile.site")' >> /etc/littler.r \
	&& ln -s /usr/share/doc/littler/examples/install.r /usr/local/bin/install.r \
	&& ln -s /usr/share/doc/littler/examples/install2.r /usr/local/bin/install2.r \
	&& ln -s /usr/share/doc/littler/examples/installGithub.r /usr/local/bin/installGithub.r \
	&& ln -s /usr/share/doc/littler/examples/testInstalled.r /usr/local/bin/testInstalled.r \
	&& install.r docopt \
	&& rm -rf /tmp/downloaded_packages/ /tmp/*.rds \
	&& rm -rf /var/lib/apt/lists/*
#│               │ └─c720d0acc6da Incremental Size: 853.6 MB
#│               │ └─c720d0acc6da Total Size: 1.0 GB

# this is the drd Dockerfile:

## Remain current
#RUN apt-get update -qq && apt-get dist-upgrade -y
#RUN update -qq && apt-get dist-upgrade -y

## From the Build-Depends of the Debian R package, plus subversion
## Check out R-devel
## Build and install according the standard 'recipe' I emailed/posted years ago
## Set Renviron.site to get libs from base R install
## Clean up
## -- all in one command to get a single AUFS layer
RUN apt-get update -qq && \
        apt-get install -y --no-install-recommends \
                bash-completion \
                bison \
                debhelper \
                default-jdk \
                g++ \
                gcc \
                gfortran \
                groff-base \
                libblas-dev \
                libbz2-dev \
                libcairo2-dev \
                libcurl4-openssl-dev \
                libfreetype6-dev \
    		libharfbuzz-dev \
                libjpeg-dev \
                liblapack-dev \
                liblzma-dev \
                libncurses5-dev \
                libpango1.0-dev \
                libpcre3-dev \
                libpng-dev \
                libreadline-dev \
                libtiff5-dev \
                libx11-dev \
                libxcb1-dev \
                libxdmcp-dev \
                libxt-dev \
                mpack \
                subversion \ 
                tcl8.6-dev \
                texinfo \
                texlive-base \
                texlive-fonts-recommended \
                texlive-generic-recommended \
                texlive-latex-base \
                texlive-latex-recommended \
                tk8.6-dev \
                x11proto-core-dev \
                xauth \
                xdg-utils \
                xfonts-base \
                xvfb \
                zlib1g-dev \
        && cd /tmp \
        && svn co http://svn.r-project.org/R/trunk R-devel \
        && cd /tmp/R-devel && \
                R_PAPERSIZE=letter \
                R_BATCHSAVE="--no-save --no-restore" \
                R_BROWSER=xdg-open \
                PAGER=/usr/bin/pager \
                PERL=/usr/bin/perl \
                R_UNZIPCMD=/usr/bin/unzip \
                R_ZIPCMD=/usr/bin/zip \
                R_PRINTCMD=/usr/bin/lpr \
                LIBnn=lib \
                AWK=/usr/bin/awk \
                CFLAGS=$(R CMD config CFLAGS) \
                CXXFLAGS=$(R CMD config CXXFLAGS) \
                ./configure --enable-R-shlib \
                        --without-blas \
                        --without-lapack \
                        --with-readline \
                        --without-recommended-packages \
                        --program-suffix=dev && \
                cd /tmp/R-devel && \
                make && \
                make install && \
                rm -rf /tmp/R-devel /tmp/downloaded_packages/ /tmp/*.rds \
        && echo "R_LIBS=\${R_LIBS-'/usr/local/lib/R/site-library:/usr/local/lib/R/library:/usr/lib/R/library'}" >> /usr/local/lib/R/etc/Renviron \
        && echo 'options("repos"="https://cran.rstudio.com")' >> /usr/local/lib/R/etc/Rprofile.site \
        && cd /usr/local/bin \
        && mv R Rdevel \
        && mv Rscript Rscriptdevel \
        && ln -s Rdevel RD \
        && ln -s Rscriptdevel RDscript \
        && dpkg --purge  \
                libblas-dev \
                libbz2-dev  \
                libcairo2-dev \
                libfontconfig1-dev \
                libfreetype6-dev \
                libglib2.0-dev \
    		libharfbuzz-dev \
                libjpeg-dev \
                liblapack-dev  \
                liblzma-dev \
                libncurses5-dev \
                libpango1.0-dev \
                libpcre3-dev \
                libpng12-dev \
                libreadline-dev \
                libtiff5-dev \
                libxft-dev \
                r-base-dev \
                tcl8.6-dev \
                texlive-base \
                texlive-fonts-recommended \
                texlive-generic-recommended \
                texlive-latex-base \
                texlive-latex-recommended \
                tk8.6-dev \
        && apt-get autoremove -qy
#│               │   └─4f37a6d72cbe Virtual Size: 445.7 MB
#│               │   └─4f37a6d72cbe Total Size: 1.5 GB

CMD ["R"]
#│               │     └─29635a375bec Virtual Size: 0.0 B Tags: shabbychef/rsid:latest
#│               │     └─29635a375bec Total Size: 1.5 GB Tags: shabbychef/rsid:latest

#for vim modeline: (do not edit)
# vim:nu:fdm=marker:fmr=FOLDUP,UNFOLD:cms=#%s:syn=Dockerfile:ft=Dockerfile:fo=croql
```
So somehow we do save around 0.9GB. I can confirm this results in a 'valid' image that has both `R` and `Rdevel`.

**Comment by shabbychef** [comment]
Via this [collection of Dockerfiles](https://github.com/shabbychef/docker-R), I can get image sizes down to 1.3G for 3.2.1 through 3.2.4 tagged versions of R from the SVN repo, but this is minus `r` or any CRAN packages, so hardly a fair comparison.

**Comment by 1beb** [comment]
@eddelbuettel unfortunate choice of acronym for that image: http://www.urbandictionary.com/define.php?term=drd

**Comment by eddelbuettel** [comment]
FWIW I just update rocker/r-base to R 3.1.4-revised, and it comes a touch over 1gb.  So the days of a drd image under 1gb are clearly gone. Question is how much space we expect all the build tools to take, and how much residual space the R-devel tree should take.  Part of me wants to think 'not more than 100mb' but i don't really have time to dig around ...

**Comment by shabbychef** [comment]
I was hoping to get a minimal "R-only" image down to, say, 300Mb. I'll poke around some more, but it does seem doubtful.

**Comment by eddelbuettel** [comment]
We could maybe do a minimal run-time without development tools (ie never install a package from source, useful?), without latex, maybe without all x11 fonts, ...    In essence start from a bare-bones Debian testing (or current Ubuntu), install r-base-core and r-recommended and purge as much as you like. 

Will it be useful?

**Comment by shabbychef** [comment]
For my purposes I want to CRAN check my packages using old R, R, and R-devel on my machine within docker containers. My download speeds at home are atrocious, so the smaller the better.

For broader acceptance, I think a minimal R image fits more with the philosophy of containers providing a single service. A minimal R image could also be the building block for images with the extras, of course, depending on user's tastes.

**Comment by eddelbuettel** [comment]
Think this through (and kudos on the initial Docker container for `R CMD check` -- I almost emailed you).  
We do not compilers, libraries, -dev packages, ... or else `R CMD check` will work only for trivial packages.

I am working a little on the opposite: a super-fat container with as build-dependencies as I can identify so that I can test reverse dependencies (now over 600 ....) for Rcpp.  I still do that serially; Docker may help me to do multiple tests in parallel (and not require everything to be installed on the host as I currently do).

**Comment by shabbychef** [comment]
FWIW, here is a [list of installed packages](https://gist.github.com/anonymous/384dceca46627f6ec411) in R 3.2.4 'minimal' image. It's hard for me to figure out which of these are truly needed, but it just feels _wrong_ that a minimal R image would have Perl, Python and Java installed.

**Comment by eddelbuettel** [comment]
The world is your oyster. `dpkg --purge ....` away as you can break are your own chains^Hcontainers.

**Comment by shabbychef** [comment]
By purging packages, I am able to get 'working' R versions down to around 700M, although docker hub reports them as ~300M. I pushed R versions 3.2.1 through 3.2.4 to [docker hub](https://hub.docker.com/r/shabbychef/rminimal/tags/). As Dirk presaged, this likely a Pyrrhic victory, since to do anything useful with these, you will likely have to _add lots of packages back_. For example, to use these images to check packages, you might need a LaTeX install and pandoc and so on. (I also tried building an image from scratch using Alpine linux, but was, again, stymied by LaTeX.)

**Comment by eddelbuettel** [comment]
I think that is the case.  I just fired up a (minimal) debian:unstable container (listed at about 120mb) and just installing r-cran-littler to get R (as well as install.r to help me install a package claiming to have a bug with Rcpp now) adds the following:

> 4 upgraded, 250 newly installed, 0 to remove and 10 not upgraded.
Need to get 248 MB of archives.
After this operation, 1030 MB of additional disk space will be used.

These simply will not be lightweight containers.  It is an additional 1gb just for _run-times_ (and maybe a few dependencies we have included for common-sense) given the minimal dependency graph computed by apt.  I think we can close this.

**Comment by shabbychef** [comment]
Arguably this was a silly issue because it could never be completely satisfied. As a concrete goal, however, I think now I would settle for an image of 1.5G or smaller. It is claimed that `drd` is on the order of [660M or so](https://hub.docker.com/r/rocker/drd/tags/), though this is by the 'funny math' of docker hub. So maybe it should be closed.
